docs/dev/writing-code.md

```diff
@@ -192,18 +201,195 @@ TODO: document when to use future vs callback
+of Python objects alive via `Py_INCREF(x)` (TODO: remove needless complexity).
+You can't always look at existing code to see "the right way" of doing things.
+
+## A More Complex Example
+
+The sample above is simplified, it only shows Python calling into C.
+But `aws_event_loop_group` has a callback that fires when it finishes shutting down.
+That means C needs to call into Python AFTER the Python `EventLoopGroup` object has been cleaned up.
+
+For C to call into Python, it must reference a Python object
+(the function itself, or an object with a member function).
+This means our binding needs to store a strong reference and
+keep that Python object alive until the callback has fired.
+
+### The Wrong Way to Build it
+
+You MUST NOT create a reference cycle!
+
+You might be tempted to give `event_loop_group_binding` a strong reference to the `EventLoopGroup` instance.
+Then C could simply call a private member function like `EventLoopGroup._on_shutdown_complete()`.
+But this design creates a reference cycle (see image below):
+
+![Diagram of bad binding with reference cycle](./binding-bad-cycle.svg)
+
+### Option 1 - Pass Callbacks to C
+
+Most of our bindings work like this:
+
+![Diagram where C References callbacks](binding-callbacks.svg)
+
+Within `EventLoopGroup.__init__()` a "callable" is defined and passed down to C.
+`event_loop_group_binding` keeps a strong reference to this Python object.
+
+When the final shutdown callback happens in C, the Python callable
+is invoked, and then the reference is released via `Py_DECREF(x)`.
+
+Destruction goes like this:
+*   When the Python code has no references to `elg`, the `EventLoopGroup` instance...
+*   The garbage collector cleans up the `EventLoopGroup`,
+    and the `PyCapsule` referenced by `EventLoopGroup._binding`.
+*   The `PyCapsule` runs its destructor function.
+    *   The destructor function calls `aws_event_loop_group_release(binding->native)`,
+        but doesn't delete the `event_loop_group_binding` struct yet.
+*   The `aws_event_loop_group` won't shutdown until nothing else is referencing it.
+    Even when the final reference is released, it still needs to wait for the threads
+    in its thread-pool to finish their shutdown process.
+*   Finally, shutdown completes and the C callback is invoked.
+    *   The C callback invokes the Python `callable`, then releases it via `Py_DECREF(x)`.
+        *   Now the garbage collector can clean up the `callable` object.
+    *   The C callback finally deletes the `event_loop_group_binding`.
+        This struct only existed to keep two strong references,
+        but now they've both been released.
+
+### Option 2 - Private Core Class
+
+Another option is to build a private `_Core` class that contains anything
+that may need to outlive the main Python object.
+This technique hasn't actually been used, but the author of this doc
+thinks it might be a graceful way to build in the future:
```

Comment:
Question: In this case, would there be two objects/classes that are managed by a bigger class that contains references to both? I'm just curious, as it sounds interesting, but I'm not totally sure I understand how it would work based on the picture...

Reply:
elaborated